students/JeroenSweerts/Lesson10/mailroom6.py

```diff
@@ -0,0 +1,209 @@
+
+    def add_donor(self, donor):
+        """Add a new donor to the collection."""
+        self.donors[donor.name] = donor.donations
+
+    def donate(self, name, donation):
+        """Add a donation to the list of donations for a specific donor.
+        If the donor doesn't exist yet, then add it to the donor collection first."""
+        if name not in self.donors.keys():
+            new_donor = Donor(name)
+            self.add_donor(new_donor)
+        if not donation.isdigit():
+            raise Exception("ERROR: You didn't enter a number or a negative number.")
+        else:
+            self.donors[name].append(int(donation))
+
+    def challenge(self, simulate, factor, min = 100, max = 9999999999999999):
+        """A philantropist can multiply the donations in the database with a indicated
+        factor. The philantropist can put a cap and floor on the donations that he wants
+        to multiply. He also can first request a simulation of what his total donation
+        would be."""
+        total = 0
+        if not str(factor).isdigit():
+            raise Exception("ERROR: You didn't enter a number or a negative number.")
+        else:
+            for key in self.donors:
+                filter_list1 = list(filter(lambda x: x > min and x < max, self.donors[key]))
+                filter_list2 = list(filter(lambda x: x <= min or x >= max, self.donors[key]))
+                if simulate == 0:
+                    self.donors[key] = list(map(lambda x: x*factor, filter_list1)) + filter_list2
+                else:
+                    M_list = list(map(lambda x: x*factor, filter_list1))
+                    if M_list:
+                        total = total + reduce(lambda x, y: x+y, M_list)
+        if simulate == 1:
+            print('total simulated donation is: '+str(total))
+
+
+
+    def createfile(self, name):
+        """create .txt file containing thank you letter for each person who made a donation"""
+        d = {'name':name, 'donation':self.donors[name][-1]}
+        outfile = open(name+'.txt', 'w')
+        outfile.write("Dear {name} , \n\n".format(**d))
+        outfile.write("\t Thank you very much for your generous donation of ${donation}. \n\n".format(**d))
+        outfile.write("\t It will be put to very good use. \n\n")
+        outfile.write("\t\t Sincerely, \n")
+        outfile.write("\t\t   -The Team \n")
+        outfile.close()
+
+    def create_report(self):
+        """For all donors in the Donor_Collection class, write a report containing
+        the donor's name, how much he gave in total, how many gifts the donor made and
+        how much the donor gave on average."""
+        print(f'{"Donor Name":20s} {"|  Total Given":20s} {"|  Num Gifts  |":20s} {"Average Gift":20s}')
+        print(f'{"-"*76}')
+        [print(f'{name:20s} ${sum(self.donors[name]):20.2f} {len(self.donors[name]):13d}${sum(self.donors[name])/len(self.donors[name]):20.2f}') for name in self.donors.keys()]
```

Comment:
comprehensions are not the best choice here since you're not capturing the output.